lib/Optimizer/Transforms/MemToReg.cpp

```diff
@@ -359,150 +297,129 @@ class RegionDataFlow {
+    return unsafeAddLiveInToBlock(block, mr);
   }
 
-  /// Add \p mr to the set of live-in definitions for \p block. This can only be
-  /// used if the parent is a function.
-  std::pair<SSAReg, bool> addLiveInToBlock(Block *block, MemRef mr) {
-    assert(block && mr && isFunctionBlock(block));
-    if (!liveInMap.count(block))
-      liveInMap.insert({block, DenseMap<MemRef, SSAReg>{}});
+  SSAReg maybeAddLiveInToBlock(Block *block, MemRef mr) {
+    assert(block && liveInMap.count(block) && mr);
     if (liveInMap[block].count(mr))
-      return {liveInMap[block][mr], /*changed=*/false};
+      return liveInMap[block][mr];
+    return addLiveInToBlock(block, mr);
+  }
 
-    // `mr` has not already been added.
-    // Add it as an argument to `block`.
-    auto ty = dereferencedType(mr.getType());
-    SSAReg newArg = block->addArgument(ty, mr.getLoc());
-    liveInMap[block][mr] = newArg;
-
-    // For each predecessor, add a load and forward the value to `block`.
-    for (auto *pred : block->getPredecessors()) {
-      for (auto iter : llvm::enumerate(pred->getSuccessors())) {
-        auto idx = iter.index();
-        auto *succ = iter.value();
-        if (succ != block)
-          continue;
-        // Create the re-load.
-        OpBuilder builder(pred->getTerminator());
-        auto sr = reloadMemoryReference(builder, mr);
-        // Update the branch's successor operands list.
-        auto branch = cast<BranchOpInterface>(pred->getTerminator());
-        branch.getSuccessorOperands(idx).append(ArrayRef<Value>{sr});
-        assert(branch.getSuccessorOperands(idx).size() ==
-               block->getNumArguments());
+  /// Record the memory reference \p mr as live-in to \p block. The live-in
+  /// value is specified as \p val. Consequently, \p val \em{must dominate} \p
+  /// block.
+  void addLiveInToBlock(Block *block, MemRef mr, SSAReg val) {
+    assert(block && liveInMap.count(block) && mr && val &&
+           !liveInMap[block].count(mr) && !isFunctionEntryBlock(block));
+    liveInMap[block][mr] = val;
+  }
+
+  /// Returns a vector of memory references. These memory references are the
+  /// ordered list of arguments to \p block.
+  SmallVector<MemRef> getLiveInToBlock(Block *block) {
+    assert(block && liveInMap.count(block));
+    std::map<unsigned, MemRef> sortedMap;
+    for (auto [mr, val] : liveInMap[block])
+      if (auto arg = dyn_cast<BlockArgument>(val))
+        if (arg.getOwner() == block)
+          sortedMap[arg.getArgNumber()] = mr;
+
+#ifndef NDEBUG
+    // Sanity check that these arguments are contiguous.
+    if (!sortedMap.empty()) {
+      auto iter = sortedMap.begin();
+      unsigned index = iter->first;
+      for (++iter; iter != sortedMap.end(); ++iter) {
+        assert(iter->first == index + 1);
+        index = iter->first;
       }
     }
-    return {newArg, /*changed=*/true};
+#endif
+
+    SmallVector<MemRef> result;
+    for (auto [index, mr] : sortedMap)
+      result.push_back(mr);
```

Comment:
These are actually trickier than they look. Since this is a return value of type `SmallVector` the vector's storage is pre-allocated on the caller's side. The size is some compiler computed constant that approximates a cache line IIRC.

If we use the size_t constructor here, we get a different flavor of `SmallVector` even though it will statically type check. So when we return to the caller from here, the content of the vector will be garbage and we'll get a sigsegv.

The old school way of doing this is to write something like
```c++
   SmallVector<T, 10> buffer;
   foo(buffer);
```
in the caller. On the callee side, we'd have
```c++
void foo(SmallVectorImpl<T> &buffer) {
  for (...)
    buffer.push_back(...);
}
```